In [1]:
import pandas as pd
import numpy as np

In [3]:
import os

os.listdir('/content')

['.config', 'sample_data']

In [4]:
from google.colab import drive

drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
import os

os.listdir('/content/drive/MyDrive')

['IMG_20241211_092923.jpg',
 'Colab Notebooks',
 'Student_Declaration_Form_akash[1][1].docx',
 'inbound111876960796203468.webp',
 'Give this in pdf way.gdoc',
 'Add 2nd photo in first photo behind the virat phot....gdoc',
 'Blood Donor App 1.0.zip',
 'app-release[1] (1).apk',
 'app-release[1].apk',
 'Portfolio.gsite',
 'Create a photo using effect and everything from 1....gdoc',
 'Akash_B_Data_Analyst_Resume_One_Page (1).gdoc',
 'Akash_B_Data_Analyst_Resume_One_Page.gdoc',
 'Akash_B_Data_Analyst_Resume_COMPLETE_1_PAGE.pdf',
 'Copy of SQL Interview Preparation.pdf']

In [7]:
from google.colab import files

uploaded = files.upload()

Saving archive (1).zip to archive (1).zip


In [8]:
import zipfile
import os

with zipfile.ZipFile('/content/archive (1).zip', 'r') as zip_ref:
    zip_ref.extractall('/content/task3_data')

In [9]:
os.listdir('/content/task3_data')

['my_file (1).csv']

In [10]:
os.listdir('/content/task3_data')

['my_file (1).csv']

In [11]:
import pandas as pd
import numpy as np

df = pd.read_csv('/content/task3_data/my_file (1).csv')

df.head()

,Rank,Peak,All Time Peak,Actual gross,Adjusted gross (in 2022 dollars),Artist,Tour title,Year(s),Shows,Average gross,Ref.
0,1,1,2,"$780,000,000","$780,000,000",Taylor Swift,The Eras Tour †,2023–2024,56,"$13,928,571",[1]
1,2,1,7[2],"$579,800,000","$579,800,000",Beyoncé,Renaissance World Tour,2023,56,"$10,353,571",[3]
2,3,1[4],2[5],"$411,000,000","$560,622,615",Madonna,Sticky & Sweet Tour ‡[4][a],2008–2009,85,"$4,835,294",[6]
3,4,2[7],10[7],"$397,300,000","$454,751,555",Pink,Beautiful Trauma World Tour,2018–2019,156,"$2,546,795",[7]
4,5,2[4],NaN,"$345,675,146","$402,844,849",Taylor Swift,Reputation Stadium Tour,2018,53,"$6,522,173",[8]


In [12]:
df.shape

(20, 11)

In [13]:
df.head()

,Rank,Peak,All Time Peak,Actual gross,Adjusted gross (in 2022 dollars),Artist,Tour title,Year(s),Shows,Average gross,Ref.
0,1,1,2,"$780,000,000","$780,000,000",Taylor Swift,The Eras Tour †,2023–2024,56,"$13,928,571",[1]
1,2,1,7[2],"$579,800,000","$579,800,000",Beyoncé,Renaissance World Tour,2023,56,"$10,353,571",[3]
2,3,1[4],2[5],"$411,000,000","$560,622,615",Madonna,Sticky & Sweet Tour ‡[4][a],2008–2009,85,"$4,835,294",[6]
3,4,2[7],10[7],"$397,300,000","$454,751,555",Pink,Beautiful Trauma World Tour,2018–2019,156,"$2,546,795",[7]
4,5,2[4],NaN,"$345,675,146","$402,844,849",Taylor Swift,Reputation Stadium Tour,2018,53,"$6,522,173",[8]


In [15]:
df.shape

(20, 11)

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 11 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   Rank                              20 non-null     int64 
 1   Peak                              9 non-null      object
 2   All Time Peak                     6 non-null      object
 3   Actual gross                      20 non-null     object
 4   Adjusted gross (in 2022 dollars)  20 non-null     object
 5   Artist                            20 non-null     object
 6   Tour title                        20 non-null     object
 7   Year(s)                           20 non-null     object
 8   Shows                             20 non-null     int64 
 9   Average gross                     20 non-null     object
 10  Ref.                              20 non-null     object
dtypes: int64(2), object(9)
memory usage: 1.8+ KB


In [17]:
df.isnull().sum()

,0
Rank,0
Peak,11
All Time Peak,14
Actual gross,0
Adjusted gross (in 2022 dollars),0
Artist,0
Tour title,0
Year(s),0
Shows,0
Average gross,0


In [18]:
df.duplicated().sum()

np.int64(0)

In [19]:
df.shape

(20, 11)

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 11 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   Rank                              20 non-null     int64 
 1   Peak                              9 non-null      object
 2   All Time Peak                     6 non-null      object
 3   Actual gross                      20 non-null     object
 4   Adjusted gross (in 2022 dollars)  20 non-null     object
 5   Artist                            20 non-null     object
 6   Tour title                        20 non-null     object
 7   Year(s)                           20 non-null     object
 8   Shows                             20 non-null     int64 
 9   Average gross                     20 non-null     object
 10  Ref.                              20 non-null     object
dtypes: int64(2), object(9)
memory usage: 1.8+ KB


In [21]:
df.isnull().sum()

,0
Rank,0
Peak,11
All Time Peak,14
Actual gross,0
Adjusted gross (in 2022 dollars),0
Artist,0
Tour title,0
Year(s),0
Shows,0
Average gross,0


In [24]:
df.describe()

,Rank,Shows
count,20.000000,20.000000
mean,10.450000,110.000000
std,5.942488,66.507617
min,1.000000,41.000000
25%,5.750000,59.000000
50%,10.500000,87.000000
75%,15.250000,134.500000
max,20.000000,325.000000


In [23]:
df.duplicated().sum()


np.int64(0)

### Observation

The dataset contains **20 rows and 11 columns**. There are **no duplicate rows** in the dataset. Missing values are present in the `Peak` column (11 missing values) and the `All Time Peak` column (14 missing values). The dataset currently contains **2 integer columns and 9 object columns**. The `Rank` column ranges from 1 to 20, while the `Shows` column ranges from 41 to 325. Several columns containing monetary values are stored as text (`object`) and will require data type correction during the cleaning process. Overall, the dataset requires missing-value handling, standardization, and data type correction before it can be used for analysis.


In [25]:
df[df['Peak'].isnull() | df['All Time Peak'].isnull()]

,Rank,Peak,All Time Peak,Actual gross,Adjusted gross (in 2022 dollars),Artist,Tour title,Year(s),Shows,Average gross,Ref.
4,5,2[4],NaN,"$345,675,146","$402,844,849",Taylor Swift,Reputation Stadium Tour,2018,53,"$6,522,173",[8]
6,7,2[10],NaN,"$280,000,000","$381,932,682",Celine Dion,Taking Chances World Tour,2008–2009,131,"$2,137,405",[11]
7,7,NaN,NaN,"$257,600,000","$257,600,000",Pink,Summer Carnival †,2023–2024,41,"$6,282,927",[12]
8,9,NaN,NaN,"$256,084,556","$312,258,401",Beyoncé,The Formation World Tour,2016,49,"$5,226,215",[13]
9,10,NaN,NaN,"$250,400,000","$309,141,878",Taylor Swift,The 1989 World Tour,2015,85,"$2,945,882",[14]
10,11,NaN,NaN,"$229,100,000[b]","$283,202,896",Beyoncé,The Mrs. Carter Show World Tour,2013–2014,132,"$1,735,606",[15][16]
11,12,NaN,14[17],"$227,400,000","$295,301,479",Lady Gaga,The Monster Ball Tour *,2009–2011,203,"$1,118,227",[18]
12,13,NaN,NaN,"$204,000,000","$251,856,802",Katy Perry,Prismatic World Tour,2014–2015,151,"$1,350,993",[19]
13,14,1[20],NaN,"$200,000,000","$299,676,265",Cher,Living Proof: The Farewell Tour ‡[21][a],2002–2005,325,"$615,385",[20]
14,15,2[c],NaN,"$194,000,000","$281,617,035",Madonna,Confessions Tour,2006,60,"$3,233,333",[5]


### Missing Data Handling

The `Peak` column contains 11 missing values and the `All Time Peak` column contains 14 missing values. These columns represent ranking information, so mean, median, or mode imputation would create artificial ranking values that are not supported by the dataset. Therefore, the missing values are retained as `NaN` rather than assigning inaccurate values. The remaining information in these rows is still useful for analysis.


In [26]:
df['Peak'] = df['Peak'].astype('string').str.replace(r'\[.*?\]', '', regex=True)

In [27]:
df['Peak'].head(10)

,Peak
0,1
1,1
2,1
3,2
4,2
5,2
6,2
7,<NA>
8,<NA>
9,<NA>


In [28]:
df['Peak'] = pd.to_numeric(df['Peak'], errors='coerce')

In [29]:
df['Peak'].dtype

Int64Dtype()

In [30]:
df['All Time Peak'] = df['All Time Peak'].astype('string').str.replace(r'\[.*?\]', '', regex=True)

In [31]:
df['All Time Peak'].head(10)

,All Time Peak
0,2
1,7
2,2
3,10
4,<NA>
5,10
6,<NA>
7,<NA>
8,<NA>
9,<NA>


In [32]:
df['All Time Peak'] = pd.to_numeric(df['All Time Peak'], errors='coerce')

In [33]:
df['All Time Peak'].dtype

Int64Dtype()

In [35]:
df.columns.tolist()

['Rank',
 'Peak',
 'All Time Peak',
 'Actual\xa0gross',
 'Adjusted\xa0gross (in 2022 dollars)',
 'Artist',
 'Tour title',
 'Year(s)',
 'Shows',
 'Average gross',
 'Ref.']

In [36]:
df['Actual\xa0gross'] = (
    df['Actual\xa0gross']
    .astype('string')
    .str.replace(r'\[.*?\]', '', regex=True)
    .str.replace('$', '', regex=False)
    .str.replace(',', '', regex=False)
)

In [37]:
df['Actual\xa0gross'].head(10)

,Actual gross
0,780000000
1,579800000
2,411000000
3,397300000
4,345675146
5,305158363
6,280000000
7,257600000
8,256084556
9,250400000


In [38]:
df['Actual\xa0gross'] = pd.to_numeric(df['Actual\xa0gross'], errors='coerce')

In [39]:
df['Actual\xa0gross'].dtype

Int64Dtype()

In [40]:
df['Adjusted\xa0gross (in 2022 dollars)'] = (
    df['Adjusted\xa0gross (in 2022 dollars)']
    .astype('string')
    .str.replace(r'\[.*?\]', '', regex=True)
    .str.replace('$', '', regex=False)
    .str.replace(',', '', regex=False)
)

In [41]:
df['Adjusted\xa0gross (in 2022 dollars)'].head(10)

,Adjusted gross (in 2022 dollars)
0,780000000
1,579800000
2,560622615
3,454751555
4,402844849
5,388978496
6,381932682
7,257600000
8,312258401
9,309141878


In [42]:
df['Adjusted\xa0gross (in 2022 dollars)'] = pd.to_numeric(
    df['Adjusted\xa0gross (in 2022 dollars)'],
    errors='coerce'
)

In [43]:
df['Adjusted\xa0gross (in 2022 dollars)'].dtype

Int64Dtype()

In [44]:
df['Average gross'] = (
    df['Average gross']
    .astype('string')
    .str.replace(r'\[.*?\]', '', regex=True)
    .str.replace('$', '', regex=False)
    .str.replace(',', '', regex=False)
)

In [45]:
df['Average gross'].head(10)

,Average gross
0,13928571
1,10353571
2,4835294
3,2546795
4,6522173
5,3467709
6,2137405
7,6282927
8,5226215
9,2945882


In [46]:
df['Average gross'] = pd.to_numeric(df['Average gross'], errors='coerce')

In [47]:
df['Average gross'].dtype

Int64Dtype()

In [48]:
df['Year(s)'].head(10)

,Year(s)
0,2023–2024
1,2023
2,2008–2009
3,2018–2019
4,2018
5,2012
6,2008–2009
7,2023–2024
8,2016
9,2015


### Observation

The dataset was standardised by removing unnecessary spaces from text fields and removing reference markers, currency symbols, and commas from numerical fields. Ranking columns were converted to numeric types, monetary columns were converted to `float64`, `Year(s)` was stored as a string because it contains both single years and year ranges, and `Ref.` was stored as a string because it contains reference information. The final data types are appropriate for analysis.


In [49]:
df['Year(s)'] = df['Year(s)'].astype('string')

In [50]:
df['Year(s)'].dtype

string[python]

In [51]:
df['Artist'] = df['Artist'].str.strip()

In [52]:
df['Artist'].head()

,Artist
0,Taylor Swift
1,Beyoncé
2,Madonna
3,Pink
4,Taylor Swift


In [53]:
df['Tour title'] = df['Tour title'].str.strip()

In [54]:
df['Ref.'] = df['Ref.'].str.strip()

In [55]:
df.dtypes

,0
Rank,int64
Peak,Int64
All Time Peak,Int64
Actual gross,Int64
Adjusted gross (in 2022 dollars),Int64
Artist,object
Tour title,object
Year(s),string[python]
Shows,int64
Average gross,Int64


In [56]:
df['Actual\xa0gross'] = df['Actual\xa0gross'].astype(float)

In [57]:
df['Adjusted\xa0gross (in 2022 dollars)'] = df['Adjusted\xa0gross (in 2022 dollars)'].astype(float)

In [58]:
df['Average gross'] = df['Average gross'].astype(float)

In [59]:
df['Ref.'] = df['Ref.'].astype('string')

In [60]:
df.dtypes

,0
Rank,int64
Peak,Int64
All Time Peak,Int64
Actual gross,float64
Adjusted gross (in 2022 dollars),float64
Artist,object
Tour title,object
Year(s),string[python]
Shows,int64
Average gross,float64


### Standardisation and Data Type Correction

The dataset was standardised and the data types were corrected for analysis. Ranking values in the `Peak` and `All Time Peak` columns were cleaned by removing reference markers such as `[4]` and converted to numeric values. Monetary columns such as `Actual gross`, `Adjusted gross`, and `Average gross` were cleaned by removing currency symbols, commas, and reference markers, and converted to `float64`. The `Year(s)` column was kept as a string because it contains both single years and year ranges. Text columns were cleaned by removing unnecessary leading and trailing spaces. The `Ref.` column was stored as a string because it contains reference information.

After correction, the dataset contains appropriate data types for further analysis.


In [61]:
Q1 = df['Shows'].quantile(0.25)
Q3 = df['Shows'].quantile(0.75)
IQR = Q3 - Q1

lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR

df[(df['Shows'] < lower) | (df['Shows'] > upper)]

,Rank,Peak,All Time Peak,Actual gross,Adjusted gross (in 2022 dollars),Artist,Tour title,Year(s),Shows,Average gross,Ref.
13,14,1,<NA>,200000000.0,299676265.0,Cher,Living Proof: The Farewell Tour ‡[21][a],2002–2005,325,615385.0,[20]


In [62]:
Q1 = df['Average gross'].quantile(0.25)
Q3 = df['Average gross'].quantile(0.75)
IQR = Q3 - Q1

lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR

df[(df['Average gross'] < lower) | (df['Average gross'] > upper)]

,Rank,Peak,All Time Peak,Actual gross,Adjusted gross (in 2022 dollars),Artist,Tour title,Year(s),Shows,Average gross,Ref.
0,1,1,2,780000000.0,780000000.0,Taylor Swift,The Eras Tour †,2023–2024,56,13928571.0,[1]
1,2,1,7,579800000.0,579800000.0,Beyoncé,Renaissance World Tour,2023,56,10353571.0,[3]


In [63]:
Q1 = df['Actual\xa0gross'].quantile(0.25)
Q3 = df['Actual\xa0gross'].quantile(0.75)
IQR = Q3 - Q1

lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR

df[(df['Actual\xa0gross'] < lower) | (df['Actual\xa0gross'] > upper)]

,Rank,Peak,All Time Peak,Actual gross,Adjusted gross (in 2022 dollars),Artist,Tour title,Year(s),Shows,Average gross,Ref.
0,1,1,2,780000000.0,780000000.0,Taylor Swift,The Eras Tour †,2023–2024,56,13928571.0,[1]
1,2,1,7,579800000.0,579800000.0,Beyoncé,Renaissance World Tour,2023,56,10353571.0,[3]


In [64]:
Q1 = df['Adjusted\xa0gross (in 2022 dollars)'].quantile(0.25)
Q3 = df['Adjusted\xa0gross (in 2022 dollars)'].quantile(0.75)
IQR = Q3 - Q1

lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR

df[(df['Adjusted\xa0gross (in 2022 dollars)'] < lower) | (df['Adjusted\xa0gross (in 2022 dollars)'] > upper)]

,Rank,Peak,All Time Peak,Actual gross,Adjusted gross (in 2022 dollars),Artist,Tour title,Year(s),Shows,Average gross,Ref.
0,1,1,2,780000000.0,780000000.0,Taylor Swift,The Eras Tour †,2023–2024,56,13928571.0,[1]


### Outlier Detection and Handling

The IQR (Interquartile Range) method was used to identify potential outliers in the numerical columns. Outliers were found in `Shows`, `Average gross`, `Actual gross`, and `Adjusted gross`. These observations represent unusually high values but are valid records in the dataset rather than obvious data-entry errors. Therefore, the identified outliers were retained and were not removed or capped.


In [65]:
df.isnull().sum()

,0
Rank,0
Peak,11
All Time Peak,14
Actual gross,0
Adjusted gross (in 2022 dollars),0
Artist,0
Tour title,0
Year(s),0
Shows,0
Average gross,0


In [66]:
df.duplicated().sum()

np.int64(0)

In [67]:
df.shape

(20, 11)

In [68]:
df.dtypes

,0
Rank,int64
Peak,Int64
All Time Peak,Int64
Actual gross,float64
Adjusted gross (in 2022 dollars),float64
Artist,object
Tour title,object
Year(s),string[python]
Shows,int64
Average gross,float64


### Before vs After Data Quality Summary

| Data Quality Check      | Before Cleaning | After Cleaning |
| ----------------------- | --------------: | -------------: |
| Total Rows              |              20 |             20 |
| Total Columns           |              11 |             11 |
| Duplicate Rows          |               0 |              0 |
| Missing `Peak`          |              11 |             11 |
| Missing `All Time Peak` |              14 |             14 |
| Monetary Data Types     |          Object |        Float64 |
| `Year(s)` Data Type     |          Object |         String |
| `Ref.` Data Type        |          Object |         String |

The dataset now has appropriate data types, standardized text fields, no duplicate rows, and cleaned numerical and monetary values. Missing ranking values were intentionally retained because they could not be reliably inferred.


### Observation

After cleaning, the dataset still contains 20 rows and 11 columns, with 0 duplicate rows. The missing ranking values remain intentionally unchanged. Monetary columns are now stored as `float64`, while `Year(s)` and `Ref.` are stored as strings. The dataset is now standardized and ready for analysis.


In [69]:
df.to_csv('/content/cleaned_tour_dataset.csv', index=False)

In [70]:
import os
os.path.exists('/content/cleaned_tour_dataset.csv')

True

### Observation

The cleaned dataset was saved as a new CSV file named `cleaned_tour_dataset.csv`. The original dataset was not overwritten. The file existence check returned `True`, confirming that the cleaned dataset was successfully saved.


### Final Observation

The dataset was successfully cleaned and prepared for analysis. Data quality was checked for missing values, duplicates, data types, and range issues. Missing ranking values were handled appropriately, duplicate records were checked, and inconsistent formatting was standardized. Monetary and other data types were corrected, and potential outliers were identified using the IQR method and retained when they represented valid records. A before-and-after quality check was completed, and the cleaned dataset was saved as a new CSV file while preserving the original data.

**Thank you!**
